### PR TITLE
Fix bug in PCPhase bind

### PR DESCRIPTION
**Pull request by comp-phys-marc**
**Context:** We are trying to see if we can run certain circuits with capture, qjit and decomposition. At the moment, this is limited by an error arising from `jax_primitives` due to a workaround which has thus far been left in place in the bind function of `PCPhase`.

**Description of the Change:** We remove the workaround and match the signature of the `PCPhase` to what is expected in `jax_primitives`.

**Benefits:** We are closer to using `PCPhase` in circuits with capture, qjit and decomps.

**Possible Drawbacks:** Not sure yet how this change will propagate.

**Related ShortCut Stories:** [sc-101673] [sc-101776]

### CODE REVIEW

**Review by JerryChen97 — CHANGES_REQUESTED** [review]
Might need sync the submodule so that we dont touch enzyme etc via this pr🤔

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -106,6 +106,10 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* Fix a bug in the bind call function for `PCPhase` where the signature did not match what was 
+  expected in `jax_primitives`.
```
Comment (JerryChen97) [inline doc/releases/changelog-dev.md]:
```suggestion
  expected in `jax_primitives`. `ctrl_qubits` was missing from positional arguments in previous signature.
```
Let's be more specific about this fix!

File: frontend/catalyst/from_plxpr/qfunc_interpreter.py [hunk 2]
```diff
@@ -355,11 +355,13 @@ def _pcphase_bind_call(*invals, op, qubits_len, params_len, ctrl_len, adjoint, h
     params_len += 1
 
     ctrl_inputs = invals[qubits_len + 2 :]
+    ctrl_wires = invals[qubits_len + 1 : -len(ctrl_inputs)]
```
Comment (JerryChen97) [inline frontend/catalyst/from_plxpr/qfunc_interpreter.py:357]: Why `ctrl_inputs` starts from `qubits_len + 2` while `ctrl_wires` starts from `qubits_len + 1` til `-len(ctrl_inputs)`? Considering that, after `ctrl_inputs` is defined, `-len(ctrl_inputs)` is exactly `qubits_len + 2`, I'll say this implies that `ctrl_wires` is literally the single index `qubits_len + 1`?
Anyway this relationship looks weird to me; haven't checked further details, but it feels like a bug...

**Review by JerryChen97 — COMMENTED** [review]
Also mind add a minimal test somewhere? Not sure if this is easy to do; curious why Catalyst's tests never caught this obvious bug...

File: frontend/catalyst/from_plxpr/qfunc_interpreter.py (hunk 2 above)
Comment (comp-phys-marc) [reply frontend/catalyst/from_plxpr/qfunc_interpreter.py:357]: After the qubits_len we take an `angle`. After that comes the `ctrl_wires` and finally the `ctrl_inputs`. It indeed seems to imply that the `ctrl_wires` are exactly one index, and in the case I have seen it is in fact true. This bind function is therefore written assuming that there is only one control wire, prior to this PR.

Comment (comp-phys-marc) [reply frontend/catalyst/from_plxpr/qfunc_interpreter.py:357]: @maliasadi can you please confirm whether there is a good reason to expect a single control here?

Comment (maliasadi) [reply frontend/catalyst/from_plxpr/qfunc_interpreter.py:357]: Yes, Looking at `jax_primitives.py` (line 1330-1331)

``` python
ctrl_qubits = qubits_or_params[qubits_len + params_len : qubits_len + params_len + ctrl_len]
ctrl_values = qubits_or_params[qubits_len + params_len + ctrl_len :]
```

With `params_len += 1` (now 2 after adding `dim`), the bind call `[*wires, angle, dim, *ctrl_wires, *ctrl_inputs]` will be correctly parsed.

File: frontend/catalyst/from_plxpr/qfunc_interpreter.py [hunk 3]
```diff
@@ -355,11 +355,13 @@ def _pcphase_bind_call(*invals, op, qubits_len, params_len, ctrl_len, adjoint, h
...
 
     return qinst_p.bind(
         *wires,
         angle,
         dim,
+        *ctrl_wires,
```
Comment (maliasadi) [inline frontend/catalyst/from_plxpr/qfunc_interpreter.py:363]: Could you add tests in `frontend/tests/pytest/from_plxpr`?

Comment (comp-phys-marc) [reply frontend/catalyst/from_plxpr/qfunc_interpreter.py:363]: I don't see any other tests there for `PCPhase` or `Controlled`. Where should I be adding a test and what should it look like?

Comment (paul0403) [reply frontend/catalyst/from_plxpr/qfunc_interpreter.py:363]: You just want a circuit execution test containing a pc phase right? You can add it to `test_capture_integration.py`, there are a bunch of circuit tests there.

Comment (paul0403) [reply frontend/catalyst/from_plxpr/qfunc_interpreter.py:363]: Remember to check the non trivial cases, e.g. PC phase in a control, adjoint, loop, etc, although these shouldn't need their own changes other than the basic pc phase bind update here.

Comment (comp-phys-marc) [reply frontend/catalyst/from_plxpr/qfunc_interpreter.py:363]: Thanks, I added `PCPhase` to the basic circuit test as well as the `Controlled` tests since my change is about the control wires being handled properly.

**Review by paul0403 — APPROVED** [review]

**Review by JerryChen97 — APPROVED** [review]
Nice!

File: frontend/test/pytest/from_plxpr/test_capture_integration.py
Comment (paul0403) [inline frontend/test/pytest/from_plxpr/test_capture_integration.py:114]: So, the `backend` argument here is a string determining the device. It will be one of `"lightning.qubit"` and `"lightning.kokkos"`. We can xfail when on kokkos.

But, since we don't want to xfail the entire original test, let's make 3 new tests, with the circuit just containing a pcphase gate 🙂

Comment (comp-phys-marc) [reply frontend/test/pytest/from_plxpr/test_capture_integration.py:114]: Would 2 tests make sense, one for PCPhase and another with the original circuit?

File: frontend/test/pytest/from_plxpr/test_capture_integration.py [hunk 4]
```diff
@@ -348,6 +350,37 @@ def circuit(theta):
...
+        capture_result = captured_circuit(theta)
+
+        qml.capture.disable()
+
+        # Capture disabled
+
+        @qml.qnode(device)
+        def circuit(theta):
+            qml.ctrl(qml.PCPhase, control=[1], control_values=[False])(theta, 2, wires=[0])
+            return qml.state()
```
Comment (paul0403) [inline frontend/test/pytest/from_plxpr/test_capture_integration.py]:
```suggestion
        @qml.qnode(device)
        def circuit(theta):
            qml.ctrl(qml.PCPhase, control=[1], control_values=[False])(theta, 2, wires=[0])
            return qml.state()
         
        capture_result = qjit(capture=True)(theta)
```
we can just do this right?

Comment (comp-phys-marc) [reply frontend/test/pytest/from_plxpr/test_capture_integration.py]: Sure

File: frontend/test/pytest/from_plxpr/test_capture_integration.py [hunk 5]
```diff
@@ -348,6 +350,33 @@ def circuit(theta):
...
 
+    @pytest.mark.parametrize("theta", (jnp.pi, 0.1, 0.0))
+    def test_ctrl_pcphase(self, backend, theta):
+        """Test the integration for a PCPhase circuit with control."""
+        if backend == "lightning.kokkos":
+            pytest.xfail(reason="PCPhase not yet implemented on Kokkos.")
+
+        device = qml.device(backend, wires=3)
+
+        # Capture enabled
```
Comment (paul0403) [inline frontend/test/pytest/from_plxpr/test_capture_integration.py:361]: I guess let's add a `qml.capture.enable()` here just to stay in line with the rest of this test file. We can deal with the `qjit(capture=True)` migration later.

File: frontend/test/pytest/from_plxpr/test_capture_integration.py [hunk 6]
```diff
@@ -348,6 +350,31 @@ def circuit(theta):
 
         assert jnp.allclose(capture_result, circuit(theta))
 
+    @pytest.mark.parametrize("theta", (jnp.pi, 0.1, 0.0))
+    def test_ctrl_pcphase(self, backend, theta):
+        """Test the integration for a PCPhase circuit with control."""
+        if backend == "lightning.kokkos":
+            pytest.xfail(reason="PCPhase not yet implemented on Kokkos.")
```
Comment (paul0403) [inline frontend/test/pytest/from_plxpr/test_capture_integration.py]:
```suggestion
            pytest.xfail(reason="Controlled PCPhase not yet implemented on Kokkos.")
```

**Review by paul0403 — APPROVED** [review]
